src/distibutions/base-distribution.ts

```diff
@@ -0,0 +1,253 @@
+import * as tc from '@actions/tool-cache';
+import * as hc from '@actions/http-client';
+import * as core from '@actions/core';
+import * as io from '@actions/io';
+
+import semver from 'semver';
+import * as assert from 'assert';
+
+import * as path from 'path';
+import os from 'os';
+import fs from 'fs';
+
+import {INodejs, INodeVersion, INodeVersionInfo} from './base-models';
+
+export default abstract class BaseDistribution {
+  protected httpClient: hc.HttpClient;
+  protected osPlat = os.platform();
+
+  constructor(protected nodeInfo: INodejs) {
+    this.httpClient = new hc.HttpClient('setup-node', [], {
+      allowRetries: true,
+      maxRetries: 3
+    });
+  }
+
+  protected abstract getDistributionUrl(): string;
+  protected abstract evaluateVersions(nodeVersions: string[]): string;
+
+  public async getNodeJsInfo() {
+    let nodeVersions: INodeVersion[] | undefined;
+    if (this.nodeInfo.checkLatest) {
+      nodeVersions = await this.getNodejsVersions();
+      const versions = this.filterVersions(nodeVersions);
+      const evaluatedVersion = this.evaluateVersions(versions);
+
+      if (evaluatedVersion) {
+        this.nodeInfo.versionSpec = evaluatedVersion;
+      }
+    }
+
+    let toolPath = this.findVersionInHoostedToolCacheDirectory();
+    if (toolPath) {
+      core.info(`Found in cache @ ${toolPath}`);
+    } else {
+      nodeVersions = nodeVersions ?? (await this.getNodejsVersions());
+      const versions = this.filterVersions(nodeVersions);
+      const evaluatedVersion = this.evaluateVersions(versions);
+      if (!evaluatedVersion) {
+        throw new Error(
+          `Unable to find Node version '${this.nodeInfo.versionSpec}' for platform ${this.osPlat} and architecture ${this.nodeInfo.arch}.`
+        );
+      }
+      const toolName = this.getNodejsDistInfo(evaluatedVersion, this.osPlat);
```

Comment:
nitpick: I think we can get the `osPlat` directly in the method since it is available as a class property. 

I am also confused by the variable name (`toolName`). We get a complex object, not just the tool name.